src/models/mSeriesChallenge.ts

```diff
@@ -0,0 +1,268 @@
+  sort: MSeriesChallengeSorting
+}
+
+export class MSeriesChallenges extends ModelList<
+MSeriesChallenge,
+MSeriesChallengeData,
+MSeriesChallengeListResponseMeta
+> {
+  constructor (mSeriesChallenges: MSeriesChallengeData[], mSeriesChallengesMeta: MSeriesChallengeListResponseMeta, sessionHandler: SessionHandler) {
+    super(MSeriesChallenge, mSeriesChallenges, mSeriesChallengesMeta, sessionHandler)
+  }
+}
+
+export class MSeriesChallenge extends Model {
+  private _mSeriesChallengeData: MSeriesChallengeData
+  private _mSeriesChallengeParticipant?: MSeriesChallengeParticipant
+
+  constructor (mSeriesChallengeData: MSeriesChallengeData, sessionHandler: SessionHandler) {
+    super(sessionHandler)
+    this._mSeriesChallengeData = mSeriesChallengeData
+  }
+
+  private setMSeriesChallenge (mSeriesChallengeData: MSeriesChallengeData) {
+    this._mSeriesChallengeData = mSeriesChallengeData
+  }
+
+  private setMSeriesChallengeParticipant (mSeriesChallengeParticipant: MSeriesChallengeParticipant) {
+    this._mSeriesChallengeParticipant = mSeriesChallengeParticipant
+  }
+
+  async reload (options: { includeOwnedParticipantData: boolean } = { includeOwnedParticipantData: false }) {
+    const { mSeriesChallenge } = await this.action('mSeriesChallenge:show', { id: this._mSeriesChallengeData.id, userId: this.sessionHandler.userId }) as MSeriesChallengeResponse
+
+    if (options.includeOwnedParticipantData) {
+      const { mSeriesChallengeParticipant } = await this.action('mSeriesChallengeParticipant:show', { mSeriesChallengeId: this._mSeriesChallengeData.id, userId: this.sessionHandler.userId }) as MSeriesChallengeParticipantResponse
+      this.setMSeriesChallengeParticipant(new MSeriesChallengeParticipant(mSeriesChallengeParticipant, this.sessionHandler))
+    }
+
+    this.setMSeriesChallenge(mSeriesChallenge)
+
+    return this
+  }
+
+  /**
+   * Method to end a challenge. Only challenge owners can end a challenge.
+   */
+  async end () {
+    await this.action('mSeriesChallenge:delete', { id: this._mSeriesChallengeData.id })
+  }
+
+  /**
+   * Method to leave a challenge. Only challenge participants can leave a challenge.
+   */
+  async leave () {
+    await this.action('mSeriesChallengeParticipant:delete', { mSeriesChallengeId: this._mSeriesChallengeData.id, userId: this.sessionHandler.userId })
+  }
+
+  /**
+   * Method to join a challenge. Challenge instance must contain joinCode to successfully join.
+   */
+  async join () {
+    if (this._mSeriesChallengeData.joinCode !== undefined) {
+      const { mSeriesChallengeParticipant } = await this.action('mSeriesChallengeParticipant:create', { joinCode: this._mSeriesChallengeData.joinCode, userId: this.sessionHandler.userId }) as MSeriesChallengeParticipantResponse
+      const p = new MSeriesChallengeParticipant(mSeriesChallengeParticipant, this.sessionHandler)
+      this.setMSeriesChallengeParticipant(p)
+      return p
+    }
+    return undefined
```

Comment:
This is bad functionality. If `join` isn't going to work when `joinCode` is undefined then this method should throw an error.

Even better, this functionality should not exist if the `MSeriesChallenge` model doesn't contain `joinCode`. (There should be 2 types of `MSeriesChallenge` models).

Reply:
I had intended to throw an error here, but I couldn't find anywhere in the SDK where we throw local errors, so I wasn't sure if that was considered bad practice in this project.

Honestly, I'm not even 100% sure this method needs to exist on this class. `joinCode` only exists in a `mSeriesChallenge` response from the API if the session user either owns or has joined the challenge. In the latter case, `join()` is pointless. So really the only time `join()` can be used is if the session user owns the challenge and has not joined it.

Reply:
Could you not query the `mSeriesChallenge` by `joinCode` first to see the challenge details and then decide to join it from a method call on that model?

Reply:
Ah yes, you are correct, that is possible. 

As far as overloading the `MSeriesChallenge` model, I'm not so sure that is the best strategy here. If this was the case, calling `getMSeriesChallenges` would return a mixed list of `MSeriesChallenge` and `MSeriesChallengeJoinable` instances. Which seems weird to me. Perhaps I'm thinking about this incorrectly though?